File: sau_mcp_server/services/account_service.py
from __future__ import annotations

from pathlib import Path

from conf import BASE_DIR


class AccountService:
    """基于本地 cookie 文件名约定，枚举各平台可用账号。"""

    def __init__(self, runtime_home: Path | None = None) -> None:
        """保存运行根目录，便于测试时注入临时目录。"""

        self.runtime_home = Path(runtime_home) if runtime_home is not None else Path(BASE_DIR)

    def list_accounts(self, platform: str) -> list[str]:
        """按平台扫描 `cookies` 目录，提取 `account_name` 列表。"""

        cookies_dir = self.runtime_home / "cookies"
        if not cookies_dir.exists():
            return []

        prefix = f"{platform}_"
        accounts: list[str] = []
        # 当前账号文件采用 `platform_account.json` 命名，这里只做最小枚举，不引入数据库依赖。
        for account_file in sorted(cookies_dir.glob(f"{platform}_*.json")):
            account_name = account_file.stem[len(prefix) :]
            if account_name:
                accounts.append(account_name)
        return accounts

    def list_accounts_by_platform(self, platforms: list[str]) -> dict[str, list[str]]:
        """批量返回平台到账号列表的映射，供能力矩阵直接消费。"""

        return {platform: self.list_accounts(platform) for platform in platforms}

File: sau_mcp_node/src/services/account-service.ts
import fs from "node:fs";
import path from "node:path";

import type Database from "better-sqlite3";

import type { PlatformName } from "../shared/platforms.js";

type AccountSnapshotRow = {
  platform: string;
  account_name: string;
  account_file: string;
  last_check_status: string | null;
  last_check_at: string | null;
};

export type AccountSummary = {
  account_name: string;
  account_file: string;
  last_check_status: string | null;
  last_check_at: string | null;
};

/**
 * 账号查询服务。
 * 先按 cookies 目录枚举账号文件，再拼上最近一次快照信息。
 */
export class AccountService {
  public constructor(
    private readonly database: Database.Database,
    private readonly baseDir: string,
  ) {}

  /**
   * 返回按平台分组的账号列表。
   */
  public listAccounts(): Record<string, AccountSummary[]> {
    const cookiesDir = path.join(this.baseDir, "cookies");
    const groupedAccounts = new Map<string, AccountSummary[]>();
    const snapshotByKey = new Map<string, AccountSnapshotRow>();
    const snapshotRows = this.database
      .prepare<[], AccountSnapshotRow>(
        `
          SELECT platform, account_name, account_file, last_check_status, last_check_at
          FROM mcp_accounts_snapshot
        `,
      )
      .all();

    for (const row of snapshotRows) {
      snapshotByKey.set(`${row.platform}:${row.account_name}`, row);
    }

    if (!fs.existsSync(cookiesDir)) {
      return {};
    }

    for (const entry of fs.readdirSync(cookiesDir, { withFileTypes: true })) {
      if (!entry.isFile() || !entry.name.endsWith(".json")) {
        continue;
      }
      const parsed = this.parseAccountFileName(entry.name);
      if (!parsed) {
        continue;
      }
      const snapshot = snapshotByKey.get(`${parsed.platform}:${parsed.accountName}`);
      const items = groupedAccounts.get(parsed.platform) ?? [];
      items.push({
        account_name: parsed.accountName,
        account_file: path.join("cookies", entry.name).replaceAll("\\", "/"),
        last_check_status: snapshot?.last_check_status ?? null,
        last_check_at: snapshot?.last_check_at ?? null,
      });
      groupedAccounts.set(parsed.platform, items);
    }

    return Object.fromEntries(
      [...groupedAccounts.entries()].map(([platform, items]) => [
        platform,
        items.sort((left, right) => left.account_name.localeCompare(right.account_name)),
      ]),
    );
  }

  /**
   * 解析账号文件名。
   * 只接受 `platform_account.json` 这一主线命名，避免把其它文件误判成账号。
   */
  private parseAccountFileName(
    fileName: string,
  ): { platform: PlatformName; accountName: string } | null {
    const withoutExtension = fileName.slice(0, -".json".length);
    const separatorIndex = withoutExtension.indexOf("_");
    if (separatorIndex <= 0) {
      return null;
    }
    const platform = withoutExtension.slice(0, separatorIndex);
    const accountName = withoutExtension.slice(separatorIndex + 1);
    if (!["douyin", "kuaishou", "xiaohongshu", "bilibili", "tencent"].includes(platform) || !accountName) {
      return null;
    }
    return { platform: platform as PlatformName, accountName };
  }
}
